### DISCUSSION TITLE
Any equivalent for our iOS friends?

### DISCUSSION THREAD

**Discussion by pablichjenkov** [bumble-tech/appyx#267] [General]
Hi guys, great job on this project. However, when evaluating its use in production the first thing that comes to my mind is the iOS team. 
We try to align the most, and using such a library like Appyx both codebases will divert in a non reversible way.
Any clue on what you do on the iOS side.
For example Square::Workflow has both Kotlin and Swift. We considered it time ago but the team decided it was too complex.

**Comment by zsoltk** [comment]
Hey @pablichjenkov , thanks! In our case iOS goes with SwiftUI navigation and remains conceptually analogous to us. However, some parts of appyx might become multiplatform in a future version still.

**Reply by pablichjenkov** [reply]
Ahh I see, honestly swiftui navigation is not much better than jetpack navigation. Both are basically navigation UI driven instead of business logic driven. I 'LL keep an eye on it.
Thanks for answering.

**Comment by dimsuz** [comment]
We have also discussed internally that it would be great to have navigation bits of Appyx decoupled from compose, to be able to share all but the UI itself in KMM projects. We had prior success with this approach while using our home-grown navigation solution and we shared everything except UI: ViewModels/Presenters/Domain/Data, everything. It worked.

Unfortunately that navigation library of ours leaves a lot to be desired in terms of features, so on our next KMM project we planned to move to something hierarchical and graph based and when I saw Appyx I was immediately excited, because it looked so much like what I've envisioned myself!

I'm soon to try it on our Compose-based Android project, but having something like this in KMM for navigation (and using Compose only on Android for rendering) would be so cool.

**Comment by zsoltk** [comment]
Hey @dimsuz! Thanks so much. As I said in some earlier comments it's an open discussion (however bland that may sound). All I can say now is stay tuned. It's not something to happen overnight, but we'll have some news on this in the future one way or another.

**Reply by dimsuz** [reply]
That's great! I fully understand that navigation is quite a complex problem. Your library is the first one actually, which I think is most close to _actually_ solving a navigation problem, because it represents it in the correct (graph + compositional) way. So please take your time, I'd say it's better to have something properly designed than something that will work for 70% of cases.

---

I sometimes think of navigation as a series of self-sufficient "flows": LoginFlow, ProfileEditFlow, MainFlow, each of which can be embedded in others (in general), so they can be hierarchically composed. And then there are these "cases":

* An event which causes navigation to go from one sibling screen to another in the same flow. This is usually what 90% of nav libs support with their push/pop/etc
* An event which causes navigation to go from screen of one deeply nested flow to the screen of another deeply nested flow. Perhaps those flows would have only root flow in common. Imagine being in an authorized state with 10 active screens and then you receive "auth token invalid" from your backend and you **have to** go all the way up to the LoginFlow and instantiate 3 new screens while also deactivating those 10 current screens and freeing their resources (DI components etc). Or another case where you receive a deeplink and also want to go in a completely other direction involving creating a new backstack

The second point is what almost any nav lib I've seen struggles with and this is where all the manner of hacks arise. Or they sacrifice compositionality, require you to define your navgraph in a single file, etc.

Sorry, got a bit carried away, but what I've tried to say is that from the looks of it, Appyx solves both of the above problems by automatically deactivating nodes which leave the graph and _providing hooks for me_ to free my resources while it does this. This is a very important detail: ability for me to provide resources cleanup from within my navigation logic, and also in a compositional way (attached to nav nodes). I'm only reading the docs, I hope to stay as excited while I'm actually trying it :)

**Comment by zsoltk** [comment]
Hey @pablichjenkov happy to say our 2.0-alpha is out and is running on Compose Multiplatform. The navigation part is also in the process of going to multiplatform, we plan to merge it soon, still as part of alpha releases. More info here: https://bumble-tech.github.io/appyx/

**Reply by pablichjenkov** [reply]
Ah great, thank you so much for letting me know. I will give it a try as soon as I get some time.
